### PR TITLE
Add digest subtypes and update dependencies

### PR DESCRIPTION
**Pull request by DemiMarie-parity**
This adds digest subtypes, so that we can ensure that different parts of the code push the correct types of digests.  I used a rather tricky macro to implement this.

I also updated dependencies and added a script for doing so.

### CODE REVIEW

File: core/sr-primitives/src/generic/digest.rs [hunk 1]
```diff
@@ -15,6 +15,59 @@
...
 //! Generic implementation of a digest.
+//!
+//! There are three k
+//! _Inherent digests_ are digests that are provided to the runtime by the
+//! consensus algorithm. They are inherent because the runtime trusts them
+//! implicitly, although the consensus algorithm does verify them.
+//!
+//! This PR passes the digests in the header to the runtime via
+//! `initialize_block`. It does _not_, however, provide any means for runtimes
+//! to actually use the information. This will be provided later.
```
Comment (bkchr) [inline core/sr-primitives/src/generic/digest.rs]: Why is here some sort of changelog? This is code documentation and not a place for a changelog.

File: core/sr-primitives/src/generic/digest.rs [hunk 2]
```diff
@@ -58,33 +111,150 @@ impl<Item> traits::Digest for Digest<Item> where
...
+		/// their own.
+		Consensus((ConsensusEngineId, Vec<u8>), (a, b)),
+		/// Any 'non-system' digest item, opaque to the native code.  Runtimes
+		/// MUST verify these, and reject any they did not produce.  These MUST
+		/// NOT be produced by native code.
+		Other((Vec<u8>), (a)),
+	}
+
+	/// A digest item that is reserved for system use. Consensus engines and
+	/// runtimes **MUST NOT** push digests in this variant. This **SHOULD** be
```
Comment (svyatonik) [inline core/sr-primitives/src/generic/digest.rs]: Currently both `AuthoritiesChange` && `ChangesTrieRoot` are generated && pushed by runtime.

**Comment by gavofyork** [comment]
any updates here? looking stale...

File: core/sr-primitives/src/generic/digest.rs (hunk 2 above)
Comment (DemiMarie-parity) [reply core/sr-primitives/src/generic/digest.rs]: I clarified that these are reserved for the SRML.

**Pull request merged**